### DISCUSSION TITLE
using multiple xcom backends in parallel

### DISCUSSION THREAD

**Discussion by diman82** [apache/airflow#23795] [Q&A] [answered]
Is it possible to add / have support for multiple custom xcom backends?
for instance, I would like to have a custom xcom backend for s3, and using this custom backend and the default one interchangeably.

**Comment by turbaszek** [comment, marked as answer]
Should be doable, for example in this [article](https://medium.com/apache-airflow/airflow-2-0-dag-authoring-redesigned-651edc397178#custom-xcom-backend) I described "type-wise" backend which defaults to native database XCom.

I would consider something like this

```py
class S3XCom:
   def serialize_value():
        # Do serialisation for S3

class PostgresXCom:
   def serialize_value():
         # Do serialisation for storing in some postgres db

# This is the backed to be used by Airflow
class CustomXComBackend(BaseXCom):

   @staticmethod
   def serialize_value(value: Any):
       if isinstance(value, [S3XCom, PostgresXCom])
           value = value.serialize_value()

        # Else, use default backend
       return BaseXCom.serialize_value(value)
```

Next in your operator when you want to use given xcom backend just return `S3XCom(some_value)` or `PostgresXCom(some_value)`. In other words create wrapper over multiple custom XCom backends.

**Comment by diman82** [comment]
@turbaszek I didn't fully unserstand the usage suggestion part - what does it  mean _just return S3XCom(some_value) or PostgresXCom(some_value)_?
I'm using xcom through the xcom_push & xcom_pull functions, which belong to the 'TaskInstance' class, so how it's possible to control what XCom backend will be used?
e.g.:
```
context['ti'].xcom_push(key='max_date', value='some_value')
max_date = context['ti'].xcom_pull(task_ids='some_task_id', key='max_date')
```

**Comment by dabla** [comment]
We have a similar question concerning a multi XCom capable backend.  We would like to be able to switch between multiple XCom backends.  For example it would be nice to be able to switch between a MinIO based XCom backend or a Persistent Volume Claim XCom backend on OpenShift or even the default one from Airflow.  Also it would be nice to be able to pull an XCom from one XCom backend and push it back to another one.  So we could define an attribute called xcom_backend on the @task decorator to specify the XCom backend we want to use, or even specify 2 dedicated attributes if we want to use a different XCom backend for the pull and the push operations.


        # here both pull and push of xcom would use the same backend:
        @task(xcom_backend="PVC") 

        # here both pull and push of xcom would use a dedicated backend:
        @task(pull_xcom_backend="PVC", push_xcom_backend="MinIO")  


Below an example of using the @task decorator which uses the "PVC" based XCom backend instead of the default Airflow one:


        @task(xcom_backend="PVC")
        def calculate_xp_per_height(pokemon_data_dict):
              """Calculates base XP per height and returns a pandas dataframe."""
        
              print(f"pokemon_data_dict: {pokemon_data_dict}")

              df = pd.DataFrame(pokemon_data_dict)

              df["xp_per_height"] = df["base_experience"] / df["height"]

              return df

I already started implementing this functionality in the Airflow source code but it required already quite some work and thus a lot of modifications (TaskInstance, BaseOperator and decorators needed to be modified already).

Maybe you guys have another idea on how we could accomplish the same functionality in another (simpler) way?
